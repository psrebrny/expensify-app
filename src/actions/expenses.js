import database from '../firebase/firebase';

export const addExpense = (expenseData = {}) => {
  return (dispatch) => {
    return database.ref('expenses').push(expenseData).then((ref) => {
      return dispatch({
        type: 'ADD_EXPENSE',
        payload: {
          ...expenseData,
          id: ref.key
        }
      });
    });
  };
};

export const removeExpense = (id = '') => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      return dispatch({
        type: 'REMOVE_EXPENSE',
        payload: {id}
      });
    });
    
  };
  
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  payload: {id, updates}
});

export const getExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      console.log(snapshot.val());
      snapshot.forEach((childSnapShot) => {
        expenses.push({
          ...childSnapShot.val(),
          id: childSnapShot.key
        });
      });
      
      return dispatch({
        type: 'GET_EXPENSES',
        payload: expenses
      });
    });
  };
};