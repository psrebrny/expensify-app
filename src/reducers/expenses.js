// Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.payload
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({id}) => id !== action.payload.id);
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.payload.id) {
          return {
            ...expense,
            ...action.payload.updates
          };
        } else {
          return expense;
        }
      });
    case 'GET_EXPENSES':
      return action.payload;
    default:
      return state;
  }
};