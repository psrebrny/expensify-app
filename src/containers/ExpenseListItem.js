import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseListItemClass = (props) => {
  let {description, amount, createdAt, id} = props;
  return (
      <Link className="list-item" to={`/edit/${id}`}>
        <div>
          <h3 className="list-item__title">{description}</h3>
          <span className="list-item__sub-title">{moment(createdAt).format('MMMM Do YYYY')}</span>
        </div>
        <div className="list-item__right-align">
          <h3 className="list-item__data">{numeral(amount).format('$0,0.00')}</h3>
          <button className="button button--secondary"
                  onClick={(e) => {
                    e.preventDefault();
                    props.removeExpense(id);
                  }}>Remove
          </button>
        </div>
      </Link>
  
  );
};

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({
    removeExpense
  }, dispatch);
};

export const ExpenseListItem = connect(null, mapDispatchToProps)(ExpenseListItemClass);