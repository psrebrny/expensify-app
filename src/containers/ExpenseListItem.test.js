import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseListItemClass } from './ExpenseListItem';
import expenses from '../test/expensesMock';

describe('containers/ExpensesListItem.js', () => {
  test('render', () => {
    const wrapper = shallow(<ExpenseListItemClass {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
  });
  
  test('render empty', () => {
    const wrapper = shallow(<ExpenseListItemClass/>);
    expect(wrapper).toMatchSnapshot();
  });
});