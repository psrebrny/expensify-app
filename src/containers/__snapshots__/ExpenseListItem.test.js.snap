// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/ExpensesListItem.js render 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Gum
    </h3>
    <span
      className="list-item__sub-title"
    >
      January 1st 1970
    </span>
  </div>
  <div
    className="list-item__right-align"
  >
    <h3
      className="list-item__data"
    >
      $195.00
    </h3>
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</Link>
`;

exports[`containers/ExpensesListItem.js render empty 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/undefined"
>
  <div>
    <h3
      className="list-item__title"
    />
    <span
      className="list-item__sub-title"
    >
      January 1st 1970
    </span>
  </div>
  <div
    className="list-item__right-align"
  >
    <h3
      className="list-item__data"
    >
      $0.00
    </h3>
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</Link>
`;
